pgjdbc/src/main/java/org/postgresql/util/ByteBufferByteStreamWriter.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2017, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * A {@link ByteStreamWriter} that writes a {@link ByteBuffer java.nio.ByteBuffer} to a byte array
+ * parameter.
+ */
+public class ByteBufferByteStreamWriter implements ByteStreamWriter {
+
+  private final ByteBuffer buf;
+  private final int length;
+
+  /**
+   * Construct the writer with the given {@link ByteBuffer}
+   *
+   * @param buf the buffer to use.
+   */
+  public ByteBufferByteStreamWriter(ByteBuffer buf) {
+    this.buf = buf;
+    this.length = buf.remaining();
+  }
+
+  @Override
+  public int getLength() {
+    return length;
+  }
+
+  @Override
+  public void writeTo(OutputStream stream) throws IOException {
+    // this _does_ involve some copying to a temporary buffer, but that's unavoidable
+    // as OutputStream itself only accepts single bytes or heap allocated byte arrays
```

Comment:
Does that mean we should focus on `transferTo(WritableByteChannel target)` kind of API instead of sticking with `OutputStream?

Reply:
I considered just trying to support someone setting a `ByteBuffer` as a parameter. But it has some issues.

Firstly, our internal socket isn't an NIO-based socket AFAICS, so we don't have a native writable channel to pass a byte buffer to. So we'd be reduced to doing what this class does and use a Channels wrapper, but if you look at the implementation it's just copying bytes into an intermediate buffer, exactly the same as we'd be doing if someone passes an InputStream. See comment further down for why that's not ideal.

I didn't want to go so far as to change our underlying socket implementation, but if / when we do that I think we could just accept a `ByteBuffer` as a parameter directly and then probably this interface would just be legacy.